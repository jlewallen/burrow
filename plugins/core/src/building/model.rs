use crate::library::model::*;

use std::str::FromStr;

#[derive(Clone, Serialize, Deserialize, PartialEq, ToTaggedJson, Reply, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Constructed {
    Area(AreaObservation),
}

impl TryFrom<Constructed> for Effect {
    type Error = TaggedJsonError;

    fn try_from(value: Constructed) -> std::result::Result<Self, Self::Error> {
        Ok(Self::Reply(value.to_tagged_json()?.into()))
    }
}

#[derive(Debug, Serialize, ToTaggedJson)]
#[serde(rename_all = "camelCase")]
struct EditorReply {}

impl Reply for EditorReply {}

#[derive(Default, Clone, Debug)]
pub struct QuickEdit {
    pub name: Option<String>,
    pub desc: Option<String>,
}

impl TryFrom<&EntityPtr> for QuickEdit {
    type Error = DomainError;

    fn try_from(value: &EntityPtr) -> std::result::Result<Self, Self::Error> {
        let name = Some(value.name()?);
        let desc = value.desc()?;
        Ok(Self { name, desc })
    }
}

const SEPARATOR: &str =
    "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";

impl FromStr for QuickEdit {
    type Err = DomainError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts: Vec<_> = s.split(SEPARATOR).collect();
        if parts.len() != 2 {
            return Err(DomainError::Anyhow(anyhow::anyhow!("malformed quick edit")));
        }

        let (name, desc) = match parts[..] {
            [name, desc] => (name, desc),
            _ => todo!(),
        };

        let name = Some(name.trim().to_owned());
        let desc = Some(desc.trim().to_owned());

        Ok(Self { name, desc })
    }
}

impl ToString for QuickEdit {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.name.as_deref().unwrap_or(""),
            SEPARATOR,
            self.desc.as_deref().unwrap_or("")
        )
    }
}
