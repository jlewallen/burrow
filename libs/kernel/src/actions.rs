use anyhow::Result;
use serde_json::Value;
use std::{fmt::Debug, rc::Rc};

use crate::{Audience, DomainEvent, When};

use super::{session::SessionRef, Entry, Surroundings};

pub use replies::*;

pub type ReplyResult = anyhow::Result<Effect>;

/// TODO Make generic over SessionRef, Surroundings, and Result.
pub trait Action: ToJson + Debug {
    fn is_read_only() -> bool
    where
        Self: Sized;

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult;
}

#[derive(Debug, Clone)]
pub struct Raised {
    pub key: String,
    pub audience: Audience,
    pub event: Rc<dyn DomainEvent>,
}

impl Raised {
    pub fn new(audience: Audience, key: String, event: Rc<dyn DomainEvent>) -> Self {
        Self {
            key,
            audience,
            event,
        }
    }

    pub fn has_prefix(&self, prefix: &str) -> bool {
        self.key.starts_with(prefix)
    }
}

#[derive(Debug, Clone)]
pub struct Incoming {
    pub key: String,
    pub value: serde_json::Value,
}

impl Incoming {
    pub fn new(key: String, value: serde_json::Value) -> Self {
        Self { key, value }
    }

    pub fn has_prefix(&self, prefix: &str) -> bool {
        self.key.starts_with(prefix)
    }
}

#[derive(Clone, Debug)]
pub struct Scheduling {
    pub key: String,
    pub when: When,
    pub message: serde_json::Value,
}

#[derive(Clone, Debug)]
#[non_exhaustive]
pub enum Perform {
    Living {
        living: Entry,
        action: Rc<dyn Action>, // TODO Consider making this recursive?
    },
    Chain(Rc<dyn Action>),
    Delivery(Incoming),
    Raised(Raised),
    Schedule(Scheduling),
    Ping(TracePath),
}

pub trait Performer {
    fn perform(&self, perform: Perform) -> Result<Effect>;
}

#[derive(Clone, Debug)]
#[non_exhaustive]
pub enum RevertReason {
    Mysterious,
    Deliberate(String),
}

#[derive(Clone, Debug)]
#[non_exhaustive]
pub enum Effect {
    Ok,
    Prevented,
    Nothing,
    // Revert(RevertReason),
    Reply(Rc<dyn Reply>),
    Pong(TracePath),
    // This is tempting. Right now we recursively call the performer. I'm not
    // sure this gives us in benefit, but it could come in really handy when we
    // start to dynamically alter behavior in chained actions. Leaving this as
    // it stands until I have a stronger opinion.
    // Chain(Rc<dyn Action>),
}

#[derive(Clone, Default)]
pub struct TracePath(Vec<String>);

impl TracePath {
    pub fn push(self, v: String) -> Self {
        Self(self.0.into_iter().chain([v]).collect())
    }
}

impl Debug for TracePath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let value: String = self.clone().into();
        f.write_str(&value)
    }
}

impl Into<String> for TracePath {
    fn into(self) -> String {
        self.0.join("")
    }
}

impl ToJson for Effect {
    fn to_json(&self) -> std::result::Result<Value, serde_json::Error> {
        // TODO I'll need to work on this, if not to make tests scale.
        match self {
            Effect::Reply(reply) => reply.to_json(),
            _ => todo!(),
        }
    }
}

impl<T: Reply + 'static> From<T> for Effect {
    fn from(value: T) -> Self {
        Self::Reply(Rc::new(value))
    }
}
