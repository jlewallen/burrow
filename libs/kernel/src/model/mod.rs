use anyhow::Result;
use std::{
    cell::RefCell,
    collections::HashMap,
    fmt::Debug,
    ops::{Deref, Index},
    rc::Rc,
};

pub mod base;
pub mod builder;
pub mod compare;
pub mod entity;
pub mod entity_ref;
pub mod entry;
pub mod props;
pub mod scopes;

pub use base::*;
pub use builder::*;
pub use entity::*;
pub use entity_ref::*;
pub use entry::*;
pub use props::*;
pub use scopes::*;

#[derive(Clone)]
pub struct EntityPtr(Rc<RefCell<Entity>>);

impl EntityPtr {
    pub fn new(e: Entity) -> Self {
        Self(Rc::new(RefCell::new(e)))
    }
}

impl Deref for EntityPtr {
    type Target = RefCell<Entity>;

    fn deref(&self) -> &Self::Target {
        self.0.as_ref()
    }
}

pub trait IntoEntry {
    fn to_entry(&self) -> Result<Entry, DomainError>;
}

impl IntoEntry for EntityRef {
    fn to_entry(&self) -> Result<Entry, DomainError> {
        use super::session::get_my_session;
        if !self.key().valid() {
            return Err(DomainError::InvalidKey);
        }
        get_my_session()?
            .entry(&LookupBy::Key(self.key()))?
            .ok_or(DomainError::DanglingEntity)
    }
}

use compare::{AnyChanges, CompareChanges, CompareError, Modified, Original};

pub fn any_entity_changes(
    l: AnyChanges<Option<Original>, EntityPtr>,
) -> Result<Option<Modified>, CompareError> {
    use compare::TreeDiff;

    let value_after = {
        let entity = l.after.borrow();

        serde_json::to_value(&*entity)?
    };

    let value_before: JsonValue = if let Some(original) = &l.before {
        match original {
            Original::String(s) => s.parse()?,
            Original::Json(v) => (*v).clone(),
        }
    } else {
        JsonValue::Null
    };

    let diff = TreeDiff {};

    diff.any_changes(AnyChanges {
        before: value_before,
        after: value_after,
    })
}
