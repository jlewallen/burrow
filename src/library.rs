use nom::{bytes::complete::take_while1, combinator::map, IResult};

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Described(String),
}

pub fn word(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
}

pub fn spaces(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| " \t".contains(c))(i)
}

pub fn noun(i: &str) -> IResult<&str, Item> {
    map(word, |s: &str| Item::Described(s.to_owned()))(i)
}
