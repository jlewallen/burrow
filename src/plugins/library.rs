pub mod parser {
    pub use crate::kernel::*;
    pub use nom::{
        branch::alt,
        bytes::complete::{tag, take_while1},
        character::complete::digit1,
        combinator::map,
        combinator::{map_res, recognize},
        sequence::{pair, preceded, separated_pair},
        IResult,
    };
    pub use tracing::*;

    pub fn word(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
    }

    pub fn spaces(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| " \t".contains(c))(i)
    }

    pub fn noun(i: &str) -> IResult<&str, Item> {
        map(word, |s: &str| Item::Named(s.to_owned()))(i)
    }

    pub fn unsigned_number(i: &str) -> IResult<&str, u64> {
        map_res(recognize(digit1), str::parse)(i)
    }

    pub fn gid_reference(i: &str) -> IResult<&str, Item> {
        map(preceded(tag("#"), unsigned_number), |n| {
            Item::GID(EntityGID::new(n))
        })(i)
    }

    pub fn noun_or_specific(i: &str) -> IResult<&str, Item> {
        alt((noun, gid_reference))(i)
    }

    pub fn named_place(i: &str) -> IResult<&str, Item> {
        alt((
            gid_reference,
            map(word, |s: &str| Item::Route(s.to_owned())),
        ))(i)
    }
}

pub mod model {
    pub use crate::kernel::*;
    pub use anyhow::Result;
    pub use serde::{Deserialize, Serialize};
    pub use serde_json::{json, Value};
    pub use std::{collections::HashMap, ops::Deref};
    pub use tracing::*;
}

pub mod actions {
    pub use crate::kernel::*;
    pub use crate::plugins::log_test;
    pub use crate::plugins::tools;
    pub use anyhow::Result;
    pub use tracing::*;
}
