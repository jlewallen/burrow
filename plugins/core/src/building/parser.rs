use crate::library::parser::*;

use super::actions::{
    AddScopeAction, BidirectionalDigAction, BuildAreaAction, DuplicateAction, EditAction,
    EditRawAction, LimboAction, MakeItemAction, ObliterateAction,
};

pub struct EditActionParser {}

impl ParsesActions for EditActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = alt((
            map(
                preceded(pair(tag("edit raw"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditRawAction { item }) },
            ),
            map(
                preceded(pair(tag("edit"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditAction { item }) },
            ),
        ))(i)?;

        Ok(Some(action))
    }
}

pub struct MakeItemParser {}

impl ParsesActions for MakeItemParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(
                tuple((tag("@make"), spaces, tag("item"), spaces)),
                string_literal,
            ),
            |name| MakeItemAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct DuplicateActionParser {}

impl ParsesActions for DuplicateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@duplicate"), spaces), noun_or_specific),
            |item| DuplicateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct LimboActionParser {}

impl ParsesActions for LimboActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(tag("@limbo"), |_| LimboAction {})(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ObliterateActionParser {}

impl ParsesActions for ObliterateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@obliterate"), spaces), noun_or_specific),
            |item| ObliterateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BidirectionalDigActionParser {}

impl ParsesActions for BidirectionalDigActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            tuple((
                preceded(pair(tag("@dig"), spaces), string_literal),
                preceded(pair(spaces, pair(tag("to"), spaces)), string_literal),
                preceded(pair(spaces, pair(tag("for"), spaces)), string_literal),
            )),
            |(outgoing, returning, new_area)| BidirectionalDigAction {
                outgoing: outgoing.into(),
                returning: returning.into(),
                new_area: new_area.into(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ScopeActionParser {}

impl ParsesActions for ScopeActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@scope"), spaces), camel_case_word),
            |scope_key| AddScopeAction {
                scope_key: scope_key.to_owned(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BuildAreaParser {}

impl ParsesActions for BuildAreaParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(tuple((tag("@build"), spaces)), string_literal),
            |name| BuildAreaAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}
