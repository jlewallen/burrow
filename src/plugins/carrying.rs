pub mod model {
    use crate::plugins::library::model::*;

    pub type CarryingResult = Result<DomainOutcome>;

    #[derive(Debug)]
    pub enum CarryingEvent {
        ItemHeld(EntityPtr),
        ItemDropped(EntityPtr),
    }

    impl DomainEvent for CarryingEvent {}

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Location {
        pub container: Option<LazyLoadedEntity>,
    }

    impl Scope for Location {
        fn serialize(&self) -> Result<serde_json::Value> {
            Ok(serde_json::to_value(self)?)
        }

        fn scope_key() -> &'static str {
            "location"
        }
    }

    impl Needs<std::rc::Rc<dyn Infrastructure>> for Location {
        fn supply(&mut self, infra: &std::rc::Rc<dyn Infrastructure>) -> Result<()> {
            self.container = infra.ensure_optional_entity(&self.container)?;
            Ok(())
        }
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    pub struct Containing {
        pub holding: Vec<LazyLoadedEntity>,
        pub capacity: Option<u32>,
        pub produces: HashMap<String, String>,
    }

    impl Scope for Containing {
        fn serialize(&self) -> Result<serde_json::Value> {
            Ok(serde_json::to_value(self)?)
        }

        fn scope_key() -> &'static str {
            "containing"
        }
    }

    impl Needs<std::rc::Rc<dyn Infrastructure>> for Containing {
        fn supply(&mut self, infra: &std::rc::Rc<dyn Infrastructure>) -> Result<()> {
            self.holding = self
                .holding
                .iter()
                .map(|r| infra.ensure_entity(r))
                .collect::<Result<Vec<_>>>()?;
            Ok(())
        }
    }

    impl Containing {
        pub fn start_carrying(&mut self, item: &EntityPtr) -> CarryingResult {
            self.holding.push(item.clone().into());

            Ok(DomainOutcome::Ok(vec![Box::new(CarryingEvent::ItemHeld(
                item.clone(),
            ))]))
        }

        pub fn stop_carrying(&mut self, item: &EntityPtr) -> CarryingResult {
            let before = self.holding.len();
            self.holding.retain(|i| i.key != item.borrow().key);
            let after = self.holding.len();
            if before == after {
                return Ok(DomainOutcome::Nope);
            }

            Ok(DomainOutcome::Ok(vec![Box::new(
                CarryingEvent::ItemDropped(item.clone()),
            )]))
        }
    }

    #[derive(Debug, Serialize, Deserialize, Default)]
    struct Carryable {}

    impl Scope for Carryable {
        fn serialize(&self) -> Result<serde_json::Value> {
            Ok(serde_json::to_value(self)?)
        }

        fn scope_key() -> &'static str {
            "carryable"
        }
    }

    impl Needs<std::rc::Rc<dyn Infrastructure>> for Carryable {
        fn supply(&mut self, _infra: &std::rc::Rc<dyn Infrastructure>) -> Result<()> {
            Ok(())
        }
    }

    pub fn discover(source: &Entity, entity_keys: &mut Vec<EntityKey>) -> Result<()> {
        if let Ok(containing) = source.scope::<Containing>() {
            // TODO Pretty sure this clone should be unnecessary, can we clone into from an iterator?
            entity_keys.extend(containing.holding.iter().map(|er| er.key.clone()))
        }
        Ok(())
    }
}

pub mod actions {
    use super::parser::{parse, Sentence};
    use crate::plugins::library::actions::*;

    #[derive(Debug)]
    struct HoldAction {
        item: Item,
    }

    impl Action for HoldAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, args: ActionArgs) -> ReplyResult {
            info!("hold {:?}!", self.item);

            let (_, user, area, infra) = args.clone();

            match infra.find_item(args, &self.item)? {
                Some(holding) => match tools::move_between(&area, &user, &holding)? {
                    DomainOutcome::Ok(_) => Ok(Box::new(SimpleReply::Done)),
                    DomainOutcome::Nope => Ok(Box::new(SimpleReply::NotFound)),
                },
                None => Ok(Box::new(SimpleReply::NotFound)),
            }
        }
    }

    #[derive(Debug)]
    struct DropAction {
        maybe_item: Option<Item>,
    }

    impl Action for DropAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, args: ActionArgs) -> ReplyResult {
            info!("drop {:?}!", self.maybe_item);

            let (_, user, area, infra) = args.clone();

            match &self.maybe_item {
                Some(item) => match infra.find_item(args, item)? {
                    Some(dropping) => match tools::move_between(&user, &area, &dropping)? {
                        DomainOutcome::Ok(_) => Ok(Box::new(SimpleReply::Done)),
                        DomainOutcome::Nope => Ok(Box::new(SimpleReply::NotFound)),
                    },
                    None => Ok(Box::new(SimpleReply::NotFound)),
                },
                None => Ok(Box::new(SimpleReply::NotFound)),
            }
        }
    }

    #[derive(Debug)]
    struct PutInsideAction {
        item: Item,
        vessel: Item,
    }

    impl Action for PutInsideAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, args: ActionArgs) -> ReplyResult {
            info!("put-inside {:?} -> {:?}", self.item, self.vessel);

            let (_, _user, _area, infra) = args.clone();

            match infra.find_item(args.clone(), &self.item)? {
                Some(item) => match infra.find_item(args, &self.vessel)? {
                    Some(vessel) => {
                        if tools::is_container(&vessel) {
                            let from = tools::container_of(&item)?;
                            match tools::move_between(&from, &vessel, &item)? {
                                DomainOutcome::Ok(_) => Ok(Box::new(SimpleReply::Done)),
                                DomainOutcome::Nope => Ok(Box::new(SimpleReply::NotFound)),
                            }
                        } else {
                            Ok(Box::new(SimpleReply::Impossible))
                        }
                    }
                    None => Ok(Box::new(SimpleReply::NotFound)),
                },
                None => Ok(Box::new(SimpleReply::NotFound)),
            }
        }
    }

    #[derive(Debug)]
    struct TakeOutAction {
        item: Item,
        vessel: Item,
    }

    impl Action for TakeOutAction {
        fn is_read_only() -> bool {
            false
        }

        fn perform(&self, args: ActionArgs) -> ReplyResult {
            info!("take-out {:?} -> {:?}", self.item, self.vessel);

            let (_, user, _area, infra) = args.clone();

            match infra.find_item(args.clone(), &self.vessel)? {
                Some(vessel) => {
                    if tools::is_container(&vessel) {
                        match infra.find_item(args, &self.item)? {
                            Some(item) => match tools::move_between(&vessel, &user, &item)? {
                                DomainOutcome::Ok(_) => Ok(Box::new(SimpleReply::Done)),
                                DomainOutcome::Nope => Ok(Box::new(SimpleReply::NotFound)),
                            },
                            None => Ok(Box::new(SimpleReply::NotFound)),
                        }
                    } else {
                        Ok(Box::new(SimpleReply::Impossible))
                    }
                }
                None => Ok(Box::new(SimpleReply::NotFound)),
            }
        }
    }

    pub fn evaluate(i: &str) -> EvaluationResult {
        Ok(parse(i).map(|(_, sentence)| evaluate_sentence(&sentence))?)
    }

    fn evaluate_sentence(s: &Sentence) -> Box<dyn Action> {
        match s {
            Sentence::Hold(e) => Box::new(HoldAction { item: e.clone() }),
            Sentence::Drop(e) => Box::new(DropAction {
                maybe_item: e.clone(),
            }),
            Sentence::PutInside(item, vessel) => Box::new(PutInsideAction {
                item: item.clone(),
                vessel: vessel.clone(),
            }),
            Sentence::TakeOut(item, vessel) => Box::new(TakeOutAction {
                item: item.clone(),
                vessel: vessel.clone(),
            }),
        }
    }

    #[cfg(test)]
    mod tests {
        use super::*;
        use crate::{
            domain::{BuildActionArgs, QuickThing},
            plugins::carrying::model::Containing,
        };

        #[test]
        fn it_holds_unheld_items() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build
                .ground(vec![QuickThing::Object("Cool Rake")])
                .try_into()?;

            let action = HoldAction {
                item: Item::Named("rake".into()),
            };
            let reply = action.perform(args.clone())?;
            let (_, person, area, _) = args.clone();

            assert_eq!(reply.to_json()?, SimpleReply::Done.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 1);
            assert_eq!(area.borrow().scope::<Containing>()?.holding.len(), 0);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_fails_to_hold_unknown_items() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build
                .ground(vec![QuickThing::Object("Cool Broom")])
                .try_into()?;

            let action = HoldAction {
                item: Item::Named("rake".into()),
            };
            let reply = action.perform(args.clone())?;
            let (_, person, area, _) = args.clone();

            assert_eq!(reply.to_json()?, SimpleReply::NotFound.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 0);
            assert_eq!(area.borrow().scope::<Containing>()?.holding.len(), 1);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_drops_held_items() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build
                .hands(vec![QuickThing::Object("Cool Rake")])
                .try_into()?;

            let action = DropAction {
                maybe_item: Some(Item::Named("rake".into())),
            };
            let reply = action.perform(args.clone())?;
            let (_, person, area, _) = args.clone();

            assert_eq!(reply.to_json()?, SimpleReply::Done.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 0);
            assert_eq!(area.borrow().scope::<Containing>()?.holding.len(), 1);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_fails_to_drop_unknown_items() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build
                .hands(vec![QuickThing::Object("Cool Broom")])
                .try_into()?;

            let action = DropAction {
                maybe_item: Some(Item::Named("rake".into())),
            };
            let reply = action.perform(args.clone())?;
            let (_, person, area, _) = args.clone();

            assert_eq!(reply.to_json()?, SimpleReply::NotFound.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 1);
            assert_eq!(area.borrow().scope::<Containing>()?.holding.len(), 0);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_fails_to_drop_unheld_items() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build
                .ground(vec![QuickThing::Object("Cool Broom")])
                .try_into()?;

            let action = DropAction {
                maybe_item: Some(Item::Named("rake".into())),
            };
            let reply = action.perform(args.clone())?;
            let (_, person, area, _) = args.clone();

            assert_eq!(reply.to_json()?, SimpleReply::NotFound.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 0);
            assert_eq!(area.borrow().scope::<Containing>()?.holding.len(), 1);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_fails_to_puts_item_in_non_containers() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let vessel = build.build()?.named("Not A Vessel")?.into_entity()?;
            let args: ActionArgs = build
                .hands(vec![
                    QuickThing::Object("key"),
                    QuickThing::Actual(vessel.clone()),
                ])
                .try_into()?;

            let action = PutInsideAction {
                item: Item::Named("key".to_owned()),
                vessel: Item::Named("vessel".to_owned()),
            };
            let reply = action.perform(args.clone())?;
            let (_world, person, _area, _) = args;

            insta::assert_json_snapshot!(reply.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 2);
            assert_eq!(vessel.borrow().scope::<Containing>()?.holding.len(), 0);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_puts_items_in_containers() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let vessel = build
                .build()?
                .named("Vessel")?
                .holding(&vec![])?
                .into_entity()?;
            let args: ActionArgs = build
                .hands(vec![
                    QuickThing::Object("key"),
                    QuickThing::Actual(vessel.clone()),
                ])
                .try_into()?;

            let action = PutInsideAction {
                item: Item::Named("key".to_owned()),
                vessel: Item::Named("vessel".to_owned()),
            };
            let reply = action.perform(args.clone())?;
            let (_world, person, _area, _) = args;

            insta::assert_json_snapshot!(reply.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 1);
            assert_eq!(vessel.borrow().scope::<Containing>()?.holding.len(), 1);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_takes_items_out_of_containers() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let key = build.build()?.named("Key")?.into_entity()?;
            let vessel = build
                .build()?
                .named("Vessel")?
                .holding(&vec![key])?
                .into_entity()?;
            let args: ActionArgs = build
                .hands(vec![QuickThing::Actual(vessel.clone())])
                .try_into()?;

            let action = TakeOutAction {
                item: Item::Contained(Box::new(Item::Named("key".to_owned()))),
                vessel: Item::Named("vessel".to_owned()),
            };
            let reply = action.perform(args.clone())?;
            let (_world, person, _area, _) = args;

            insta::assert_json_snapshot!(reply.to_json()?);

            assert_eq!(person.borrow().scope::<Containing>()?.holding.len(), 2);
            assert_eq!(vessel.borrow().scope::<Containing>()?.holding.len(), 0);

            build.close()?;

            Ok(())
        }
    }
}

pub mod parser {
    use crate::plugins::library::parser::*;

    #[derive(Debug, Clone, Eq, PartialEq)]
    pub enum Sentence {
        Hold(Item),
        Drop(Option<Item>),
        PutInside(Item, Item),
        TakeOut(Item, Item),
    }

    pub fn parse(i: &str) -> IResult<&str, Sentence> {
        alt((hold, drop, put_inside, take_out))(i)
    }

    fn hold(i: &str) -> IResult<&str, Sentence> {
        map(separated_pair(tag("hold"), spaces, noun), |(_, target)| {
            Sentence::Hold(target)
        })(i)
    }

    fn drop(i: &str) -> IResult<&str, Sentence> {
        let specific = map(separated_pair(tag("drop"), spaces, noun), |(_, target)| {
            Sentence::Drop(Some(target))
        });

        let everything = map(tag("drop"), |_| Sentence::Drop(None));

        alt((specific, everything))(i)
    }

    fn put_inside(i: &str) -> IResult<&str, Sentence> {
        let item = map(separated_pair(tag("put"), spaces, noun), |(_, target)| {
            target
        });

        map(
            separated_pair(separated_pair(item, spaces, tag("inside of")), spaces, noun),
            |(item, target)| Sentence::PutInside(item.0, target),
        )(i)
    }

    fn take_out(i: &str) -> IResult<&str, Sentence> {
        let item = map(separated_pair(tag("take"), spaces, noun), |(_, target)| {
            target
        });

        map(
            separated_pair(separated_pair(item, spaces, tag("out of")), spaces, noun),
            |(item, target)| Sentence::TakeOut(Item::Contained(Box::new(item.0)), target),
        )(i)
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn it_parses_hold_noun_correctly() {
            let (remaining, actual) = parse("hold rake").unwrap();
            assert_eq!(remaining, "");
            assert_eq!(actual, Sentence::Hold(Item::Named("rake".into())));
        }

        #[test]
        fn it_parses_solo_drop_correctly() {
            let (remaining, actual) = parse("drop").unwrap();
            assert_eq!(remaining, "");
            assert_eq!(actual, Sentence::Drop(None));
        }

        #[test]
        fn it_parses_drop_noun_correctly() {
            let (remaining, actual) = parse("drop rake").unwrap();
            assert_eq!(remaining, "");
            assert_eq!(actual, Sentence::Drop(Some(Item::Named("rake".into()))));
        }

        #[test]
        fn it_parses_put_x_inside_of_y() {
            let (remaining, actual) = parse("put key inside of vessel").unwrap();
            assert_eq!(remaining, "");
            assert_eq!(
                actual,
                Sentence::PutInside(Item::Named("key".into()), Item::Named("vessel".into()))
            );
        }

        #[test]
        fn it_parses_take_x_out_of_y() {
            let (remaining, actual) = parse("take key out of vessel").unwrap();
            assert_eq!(remaining, "");
            assert_eq!(
                actual,
                Sentence::TakeOut(
                    Item::Contained(Box::new(Item::Named("key".into()))),
                    Item::Named("vessel".into())
                )
            );
        }

        #[test]
        fn it_errors_on_unknown_text() {
            let actual = parse("hello");
            assert!(actual.is_err());
        }
    }
}
