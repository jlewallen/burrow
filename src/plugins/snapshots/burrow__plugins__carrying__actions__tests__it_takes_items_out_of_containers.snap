---
source: src/plugins/carrying.rs
expression: reply.to_json()?
---
"done"
